### DISCUSSION TITLE
OpenGL mouse picking with ray casting

### DISCUSSION THREAD

**Discussion by kayomarz** [kaveh808/kons-9#223] [Ideas]
Hi, I've been experimenting with kons-9 and object picking and created a branch - [opengl-mouse-picking-with-ray-casting](https://github.com/kaveh808/kons-9/tree/opengl-mouse-picking-with-ray-casting)

Would look forward to comments and suggestions.

Have also created draft pull request https://github.com/kaveh808/kons-9/pull/222 only to help compare code and get comments.

The pull request also describes how to try out the object picker, although the picking isn't very accurate yet.

**Comment by kaveh808** [comment]
Nice work. This is promising. If you can get ray geometry intersection (as opposed to ray AABB) working, that would be a great addition to the system.

**Reply by kayomarz** [reply]
Sure, will try it out.

**Reply by kayomarz** [reply]
Have implemented ray geometry intersection in the same branch. Please see this comment in the original draft PR - https://github.com/kaveh808/kons-9/pull/222#issuecomment-1606216914

**Reply by kaveh808** [reply]
Very nice. I'll review the code changes in the next few days.

**Reply by kayomarz** [reply]
I will make a few more commits over the next 3 days and ping you once its done so you can review the whole thing together. Thanks Kaveh.

**Reply by kayomarz** [reply]
@kaveh I have committed the updates and updated the [PR description](https://github.com/kaveh808/kons-9/pull/222), ready for review.

**Reply by kaveh808** [reply]
Thank you. I will review the changes soon.

**Comment by aykaramba** [comment]
amazing.

**Comment by kaveh808** [comment]
Made a small update so mouse-click will not crash when shapes do not have an intersect method.

**Reply by kayomarz** [reply]
Did it crash with the boid system / growth environment? ~~I think adding an intersect method for [shape-group](https://github.com/kaveh808/kons-9/blob/535050b0bba67e5886c2baeca88b40e0eadf5b98/src/kernel/shape-group.lisp#L6) will be an improvement.~~   It turns out that `shape-group` does not need to be tested for intersection because `find-shapes` which is being used, traverses all shapes including those in a shape-group.

**Reply by kaveh808** [reply]
It crashed when I clicked in a scene with a curve in it.

**Comment by kaveh808** [comment]
I'm wondering if we could/should implement a WebGPU back end:

https://www.reddit.com/r/lisp/comments/1552s8i/crossplatform_webgpu_from_cl/

**Reply by kayomarz** [reply]
not being up to date with GPU backends/APIs I'm not the best person to add my two cents :) Since WebGPU is designed to support  computational capabilities apart from graphics I find it an interesting technology to keep an eye on.

**Reply by kaveh808** [reply]
Good point.

**Comment by awolven** [comment]
The vulkan backend is cross platform and requires less dependencies than a
webgpu backend.  I'm a little offended that you say this yet have not used
the vulkan backend I spent six weeks coding for you.

**Reply by kaveh808** [reply]
I apologize if I dropped the ball. Integrating the Vulkan code was beyond my knowledge, and no one stepped forward to tackle it.

Personally, I'm satisfied with the simple OpenGL 1.1 functionality for now, but keep an eye on eventual enhancements. I came across an article which claimed WebGPU was the cross-platform-and-browser solution for the future. Don't know if that's the case.

**Comment by kayomarz** [comment]
I will attempt to merge the vulkan code into our main branch.  I haven't worked with Vulkan but it will be a good learning opportunity. This is a great project and I would be happy to contribute just as all of you have done so far. I will check out the earlier discussions about it and take it  on from there.

**Reply by kaveh808** [reply]
That would be very helpful. Thank you @kayomarz .

Let's make the Vulkan feature as an option alongside the existing OpenGL if possible. At least until I can become more familiar with the functionality.

**Comment by kayomarz** [comment]
@kaveh808 I tried pushing a small fix to the main branch related to object picking but the push failed due to no permissions.

Have made a PR instead: https://github.com/kaveh808/kons-9/pull/234

**Reply by kaveh808** [reply]
Done. Thanks.

**Comment by kayomarz** [comment]
With PR https://github.com/kaveh808/kons-9/pull/235 we should also be able to pick lines and curves.
